docs(proxy): add ssl guidelines for communication between Agent and Datadog when using a proxy (#14780)
DataDog/documentation: content/en/agent/proxy.md

## HAProxy

[HAProxy][1] is a free, fast, and reliable solution offering proxying for TCP and HTTP applications. While HAProxy is usually used as a load balancer to distribute incoming requests to pools servers, you can also use it to proxy Agent traffic to Datadog from hosts that have no outside connectivity :

`agent -> haproxy -> Datadog`

This is the best option if you do not have a web proxy readily available in your network, and you wish to proxy a large number of Agents. In some cases, a single HAProxy instance is sufficient to handle local Agent traffic in your network as each proxy can accommodate upwards connections of 1000 Agents. **Note**: This figure is a conservative estimate based on the performance of m3.xl instances specifically. Numerous network-related variables can influence load on proxies. As always, deploy under a watchful eye. See the [HAProxy documentation][2] for additional information.

> The communication between HAProxy and Datadog will always be encrypted and it is recommended to encrypt the communication between the Agent host and the HAProxy host if they are not part of the same local network.

Reviewer: This should probably be reflected in NGINX doc wording too

> In order to encrypt data between the Agent and HAProxy, you will need to create a Subject Alternative Name (SAN) certificate for the HAProxy host. This certificate should contain both the public certificate and private key. See this [HAProxy blog post][3] for more information.

Author: > Isn't that just a regular SSL/TLS certificate? Do we mean an x509 certificate?

Yeah, we mean an x509 certificate here but if I correctly understand, there is a field named `Common Name` (or CN). which is deprecated by the [RFC 2818](https://datatracker.ietf.org/doc/html/rfc2818#section-3.1) in 2000 and I believe you can still create certificate using `CN` instead of `SAN`.

```
If a subjectAltName extension of type dNSName is present, that MUST be used as the identity. 
...
Although the use of the Common Name is existing practice, it is deprecated 
and Certification Authorities are encouraged to use the dNSName instead
```

 I added this precision because the Agent throws an error when trying to verify a certificate with CN and asks for SAN instead. Maybe should I say `you will need to create a x509 certificate with Subject Alternative Name (SAN) extension` ?

---

> I think this can be reworded just to say that "you will need to create a Subject Alternative Name (SAN) certificate and its private key for the HAProxy host" ? 

The problem here is that NGINX has 2 attributes : 

-  `ssl_certificate`  for the certificate (usually `.crt`)
-  `ssl_certificate_key` for the key (usually `.key`)

But HAProxy has only one attribute `ssl crt <cert.pem>` where `<cert.pem>` is a file containing both the certificate and the private key (can be created with `cat certificate.crt private_key.key > cert.pem`). I wanted to emphasise the fact that the certificate and private key must be in the same file but I agree that it is unclear so feel free to give any suggestions on how to rephrase it 🙏

> In order to encrypt data between the Agent and HAProxy, you will need to create a Subject Alternative Name (SAN) certificate for the HAProxy host. This certificate should contain both the public certificate and private key. See this [HAProxy blog post][3] for more information.

Reviewer: > Maybe should I say `you will need to create a x509 certificate with Subject Alternative Name (SAN) extension`

You can't have a SAN field without it being an x509 certificate I believe, this works great - makes it clear about what they need to do. 

> feel free to give any suggestions on how to rephrase it

Maybe we can call it a certificate bundle? WDYT?
```
This certificate bundle (*.pem) should contain both the public certificate and private key. See this [HAProxy blog post][3] for more information.
```

> In order to encrypt data between the Agent and HAProxy, you will need to create a Subject Alternative Name (SAN) certificate for the HAProxy host. This certificate should contain both the public certificate and private key. See this [HAProxy blog post][3] for more information.

Author: Done here : cca6a1513774cf4d69a0b2d28a621cbba772027e

### Proxy forwarding with HAProxy

#### HAProxy configuration

HAProxy should be installed on a host that has connectivity to Datadog. Use one of the following configuration files if you do not already have it configured.

> **Note**: It is recommended to use the `HTTPS` configuration file if the Agent and HAProxy are not part of the same local network.

Reviewer: We should add a note here and in https version for users to replace all `<VARIABLE_NAME>` sections to prevent them from blindly copying them. This should probably be reflected in NGINX doc wording too.

    balance roundrobin
    mode http
    # The following configuration is for HAProxy 1.8 and newer

>     server-template mothership 5 haproxy-app.agent.{{< region-param key="dd_site" >}}:443 check port 443 ssl verify <PATH_TO_CERTIFICATES> check resolvers my-dns init-addr none resolve-prefer ipv4

Author: Do you think that keeping `ssl verify none` in the configuration and add explanation on how to verify Datadog certificate would be better ?

> This may be something to cover with a separate config type as well.

Do you mean with new tabs ?

>     server-template mothership 5 haproxy-app.agent.{{< region-param key="dd_site" >}}:443 check port 443 ssl verify <PATH_TO_CERTIFICATES> check resolvers my-dns init-addr none resolve-prefer ipv4

Reviewer: > Do you think that keeping ssl verify none in the configuration and add explanation on how to verify Datadog certificate would be better ?

To me that would be the least friction for adopters but I can't comment on it confidently since I don't know for sure what the specific intents of the effort are.

> Do you mean with new tabs ?

Just mean that we can have a config with `ssl verify none` and then one with either an explanation (like you mentioned) or a full new config that has verification of the backend.

>     server-template mothership 5 haproxy-app.agent.{{< region-param key="dd_site" >}}:443 check port 443 ssl verify <PATH_TO_CERTIFICATES> check resolvers my-dns init-addr none resolve-prefer ipv4

Author: Honestly I don't really know what to think about this part. I feel like we should prioritize 'security' and provide 'correct' guidelines (actual ones have security issues) even if it increases the setup / maintenance complexity.

But I completely understand the fact that we must keep the configuration simple. Imo the current trade-off is : 

## With the certificate verification 
- Pros :
     - More accurate guidelines / more secure
     - More awareness from customers about certificates verification (as they need to setup it)
- Cons :
     - More complexity on the setup / maintenance 

## Without verification 
- Pros : 
    -  Easy to setup
- Cons : 
    - Not optimal guidelines 
    -  Customers might not be aware of the provided configuration problems

Did I miss something ? 

My opinion is that I would let the configuration as it and add more details about how to do deal with `<PATH_TO_CERIFICATES>`  (they can still replace it by `none` as an alternative). But you have a better vision than me about what is better for customers and how you want the proxy guidelines to evolve in the future so feel free to decide.

# You do not need credentials to view this page and you can
# turn it off once you are done with setup.
listen stats
    bind *:3833
    mode http
    stats enable
    stats uri /

# This section is to reload DNS Records
# Replace <DNS_SERVER_IP> and <DNS_SECONDARY_SERVER_IP> with your DNS Server IP addresses.
# For HAProxy 1.8 and newer
resolvers my-dns
    nameserver dns1 <DNS_SERVER_IP>:53
    nameserver dns2 <DNS_SECONDARY_SERVER_IP>:53
    resolve_retries 3
    timeout resolve 2s
    timeout retry 1s
    accepted_payload_size 8192
    hold valid 10s
    hold obsolete 60s

# This declares the endpoint where your Agents connects for
# sending metrics (for example, the value of "dd_url").
frontend metrics-forwarder
    bind *:3834 ssl crt <PATH_TO_PROXY_CERTIFICATE>
    mode http
    option tcplog
    default_backend datadog-metrics

    use_backend datadog-api if { path_beg -i  /api/v1/validate }
    use_backend datadog-flare if { path_beg -i  /support/flare/ }

# This declares the endpoint where your Agents connects for
# sending traces (for example, the value of "endpoint" in the APM
# configuration section).
frontend traces-forwarder

>     bind *:3835 ssl crt <PATH_TO_PROXY_CERTIFICATE>

Reviewer: Not sure about the name here since it should also imply that it has a key with it. Maybe this?
```suggestion
    bind *:3835 ssl crt <PATH_TO_PROXY_CERTIFICATE_PEM>
```

    mode http
    option tcplog
    default_backend datadog-database-monitoring-metrics

# This declares the endpoint where your Agents connects for
# sending database monitoring samples (e.g the value of "database_monitoring.samples.dd_url")
frontend database_monitoring_samples_frontend
    bind *:3840 ssl crt <PATH_TO_PROXY_CERTIFICATE>
    mode http
    option tcplog
    default_backend datadog-database-monitoring-samples

# This declares the endpoint where your Agents connects for
# sending Network Devices Monitoring metadata (e.g the value of "network_devices.metadata.dd_url")
frontend network_devices_metadata_frontend
    bind *:3841 ssl crt <PATH_TO_PROXY_CERTIFICATE>
    mode http
    option tcplog
    default_backend datadog-network-devices-metadata

# This declares the endpoint where your Agents connects for
# sending Network Devices SNMP Traps data (e.g the value of "network_devices.snmp_traps.forwarder.dd_url")
frontend network_devices_snmp_traps_frontend
    bind *:3842 ssl crt <PATH_TO_PROXY_CERTIFICATE>
    mode http
    option tcplog
    default_backend datadog-network-devices-snmp-traps

# This declares the endpoint where your Agents connect for

> # sending Instrumentations Telemetry data (e.g. the value of "apm_config.telemetry.dd_url")

Author: ```suggestion
# sending Instrumentation Telemetry data (e.g. the value of "apm_config.telemetry.dd_url")
```

backend datadog-instrumentations-telemetry
    balance roundrobin
    mode tcp
    # The following configuration is for HAProxy 1.8 and newer
    server-template mothership 5 instrumentation-telemetry-intake.{{< region-param key="dd_site" >}}:443 check port 443 ssl verify <PATH_TO_CERTIFICATES> check resolvers my-dns init-addr none resolve-prefer ipv4
    # Uncomment the following configuration for older HAProxy versions
    # server mothership instrumentation-telemetry-intake.{{< region-param key="dd_site" >}}:443 check port 443 ssl verify <PATH_TO_CERTIFICATES>

backend datadog-appsec-events # deprecated
    balance roundrobin
    mode tcp
    # The following configuration is for HAProxy 1.8 and newer
    server-template mothership 5 appsecevts-intake.{{< region-param key="dd_site" >}}:443 check port 443 ssl verify <PATH_TO_CERTIFICATES> check resolvers my-dns init-addr none resolve-prefer ipv4
    # Uncomment the following configuration for older HAProxy versions
    # server mothership appsecevts-intake.{{< region-param key="dd_site" >}}:443 check port 443 ssl verify <PATH_TO_CERTIFICATES>
```

{{% /tab %}}
{{< /tabs >}}

> **Note**: Download the certificate with one of the following commands:

Reviewer: The download of backend certificates should probably be mentioned before the config if we keep the `PATH_TO_CERTIFICATES` in the first example.

Reviewer: ```suggestion
**Note**: Install the common root CA certificates with one of the following commands:
```

```shell
sudo apt-get install ca-certificates # (Debian, Ubuntu)
yum install ca-certificates # (CentOS, Red Hat)
```

The file might be located at `/etc/ssl/certs/ca-certificates.crt` for Debian, Ubuntu or `/etc/ssl/certs/ca-bundle.crt` for CentOS, Red Hat.

> **Note**: You can use `ssl verify none` instead of `ssl verify <PATH_TO_CERTIFICATES>` if you are unable to get the certificates on the proxy host, but be aware that HAProxy will not be able to verify Datadog intake identity.

Reviewer: ```suggestion
**Note**: You can use `ssl verify none` instead of `ssl verify <PATH_TO_CERTIFICATES>` if you are unable to get the certificates on the proxy host, but be aware that HAProxy will not be able to verify Datadog's intake certificate in that case.
```

Edit each Agent to point to HAProxy by setting its `dd_url` to the address of HAProxy, for example: `haproxy.example.com`.
This `dd_url` setting can be found in the `datadog.yaml` file.

`dd_url: <SCHEME>://haproxy.example.com:3834`

> Replace `<SCHEME>` by `https` if you want your data to be encrypted between the Agent and HAProxy, or by `http` otherwise.

Reviewer: We should add that the scheme needs to match their HAProxy settings

Author: I rephrased it here : 41ea6a4696acb87e90d0bef3aa1899e4a3ab4669 WDYT ?

logs_config:
    use_http: true
    logs_dd_url: haproxy.example.com:3838
    # Comment the line below to use encryption between the Agent and HAProxy
    logs_no_ssl: true

database_monitoring:
    metrics:
        logs_dd_url: haproxy.example.com:3839
        # Comment the line below to use encryption between the Agent and HAProxy
        logs_no_ssl: true
    activity:
        logs_dd_url: haproxy.example.com:3839
        # Comment the line below to use encryption between the Agent and HAProxy
        logs_no_ssl: true
    samples:
        logs_dd_url: haproxy.example.com:3840
        # Comment the line below to use encryption between the Agent and HAProxy
        logs_no_ssl: true

network_devices:
    metadata:
        logs_dd_url: haproxy.example.com:3841
        # Comment the line below to use encryption between the Agent and HAProxy
        logs_no_ssl: true
    snmp_traps:
        forwarder:
            logs_dd_url: haproxy.example.com:3842
            # Comment the line below to use encryption between the Agent and HAProxy
            logs_no_ssl: true

appsec_config (deprecated):
    appsec_dd_url: haproxy.example.com:3844
```

> When using encryption, if the Agent cannot access the proxy certificate or is unable to validate it, you can then edit the `datadog.yaml` Agent configuration file and set `skip_ssl_validation` to `true`.

Reviewer: ```suggestion
When using encryption between the Agent and HAProxy, if the Agent does not have access to the proxy certificate, is unable to validate it, or the validation is not needed, you can edit the `datadog.yaml` Agent configuration file and set `skip_ssl_validation` to `true`.
```

> With this option set to `true`, the Agent will skip the certificate validation step so it will not be able to verify the identity of the proxy but the communication will still be encrypted.

Reviewer: ```suggestion
With this option set to `true`, the Agent will skip the certificate validation step so it will not verify the identity of the proxy but the communication will still be encrypted with SSL/TLS.
```

## NGINX

[NGINX][4] is a web server which can also be used as a reverse proxy, load balancer, mail proxy, and HTTP cache. You can also use NGINX as a proxy for your Datadog Agents:

`agent ---> nginx ---> Datadog`

The communication between NGINX and Datadog will always be encrypted and it is recommended to encrypt the communication between the Agent host and the NGINX host if they are not part of the same local network. 
In order to encrypt data between the Agent and NGINX, you will need to create a Subject Alternative Name (SAN) certificate for the NGINX host. 

### Proxy forwarding with NGINX

#### NGINX configuration

NGINX should be installed on a host that has connectivity to Datadog. Use one of the following configuration files if you do not already have it configured.

> **Note**: It is recommended to use the `HTTPS` configuration file if the Agent and NGINX are not part of the same local network.

Reviewer: ```suggestion
**Note**: It is recommended to use the `HTTPS` configuration file if the Agent and NGINX are not part of the same local isolated network.
```

        proxy_ssl_verify on;
        proxy_ssl on;
        proxy_pass ndm-intake.{{< region-param key="dd_site" >}}:443;
    } 
    server {
        listen 3842 ssl; #listen for network devices traps
        proxy_ssl_verify on;
        proxy_ssl on;
        proxy_pass snmp-traps-intake.{{< region-param key="dd_site" >}}:443;
    }
    server {
        listen 3843 ssl; #listen for instrumentations telemetry data
        proxy_ssl_verify on;
        proxy_ssl on;
        proxy_pass instrumentation-telemetry-intake.{{< region-param key="dd_site" >}}:443;
    }
    server {
        listen 3844 ssl; #listen for appsec events (deprecated)
        proxy_ssl_verify on;
        proxy_ssl on;
        proxy_pass appsecevts-intake.{{< region-param key="dd_site" >}}:443;
    }
}
```
{{% /tab %}}
{{< /tabs >}}

**Note**: Download the certificate with one of the following commands:

```shell
sudo apt-get install ca-certificates # (Debian, Ubuntu)
yum install ca-certificates # (CentOS, Red Hat)
```

The file might be located at `/etc/ssl/certs/ca-certificates.crt` for Debian, Ubuntu or `/etc/ssl/certs/ca-bundle.crt` for CentOS, Red Hat.

> **Note**: You can remove `proxy_ssl_verify on` if you are unable to get the certificates on the proxy host, but be aware that NGINX will not be able to verify Datadog intake identity.

Author: ```suggestion
**Note**: You can remove `proxy_ssl_verify on` if you are unable to get the certificates on the proxy host, but be aware that NGINX will not be able to verify Datadog's intake certificate in that case.
```

    process_dd_url: <SCHEME>://nginx.example.com:3837

logs_config:
    use_http: true
    logs_dd_url: nginx.example.com:3838
    # Comment the line below to use encryption between the Agent and NGINX
    logs_no_ssl: true

database_monitoring:
    metrics:
        logs_dd_url: nginx.example.com:3839
        # Comment the line below to use encryption between the Agent and NGINX
        logs_no_ssl: true
    activity:
        logs_dd_url: nginx.example.com:3839
        # Comment the line below to use encryption between the Agent and NGINX
        logs_no_ssl: true
    samples:
        logs_dd_url: nginx.example.com:3840
        # Comment the line below to use encryption between the Agent and NGINX
        logs_no_ssl: true

network_devices:
    metadata:
        logs_dd_url: nginx.example.com:3841
        # Comment the line below to use encryption between the Agent and NGINX
        logs_no_ssl: true
    snmp_traps:
        forwarder:
            logs_dd_url: nginx.example.com:3842
            # Comment the line below to use encryption between the Agent and NGINX
            logs_no_ssl: true

appsec_config (deprecated):
    appsec_dd_url: haproxy.example.com:3844

```

> When using encryption, if the Agent cannot access the proxy certificate or is unable to validate it, you can then edit the `datadog.yaml` Agent configuration file and set `skip_ssl_validation` to `true`.

Author: ```suggestion
When using encryption between the Agent and NGINX, if the Agent does not have access to the proxy certificate, is unable to validate it, or the validation is not needed, you can edit the `datadog.yaml` Agent configuration file and set `skip_ssl_validation` to `true`.
```

> With this option set to `true`, the Agent will skip the certificate validation step so it will not be able to verify the identity of the proxy but the communication will still be encrypted.

Author: ```suggestion
With this option set to `true`, the Agent will skip the certificate validation step so it will not verify the identity of the proxy but the communication will still be encrypted with SSL/TLS.
```

        proxy_ssl on;
        proxy_pass dbm-metrics-intake.{{< region-param key="dd_site" >}}:443;
    }
    server {
        listen 3840; #listen for database monitoring samples
        proxy_ssl_verify on;
        proxy_ssl on;
        proxy_pass dbquery-intake.{{< region-param key="dd_site" >}}:443;
    }
    server {
        listen 3841; #listen for network devices metadata
        proxy_ssl_verify on;
        proxy_ssl on;
        proxy_pass ndm-intake.{{< region-param key="dd_site" >}}:443;
    }
    server {
        listen 3842; #listen for network devices traps
        proxy_ssl_verify on;
        proxy_ssl on;
        proxy_pass snmp-traps-intake.{{< region-param key="dd_site" >}}:443;
    }
    server {
        listen 3843; #listen for instrumentations telemetry data
        proxy_ssl_verify on;
        proxy_ssl on;
        proxy_pass instrumentation-telemetry-intake.{{< region-param key="dd_site" >}}:443;
    }
    server {    
        listen 3844; #listen for appsec events (deprecated)
        proxy_ssl_verify on;
        proxy_ssl on;
        proxy_pass appsecevts-intake.{{< region-param key="dd_site" >}}:443;
    }
}
```

{{% /tab %}}
{{% tab "HTTPS" %}}

> This configuration adds SSL/TLS encryption on communication between the Agent and NGINX. The variable `<PATH_TO_PROXY_CERTIFICATE>` should be replaced by the path to the proxy public certificate and `<PATH_TO_PROXY_CERTIFICATE_KEY>` by the path to the private key.

Reviewer 2: ```suggestion
This configuration adds SSL/TLS encryption on communication between the Agent and NGINX. Replace `<PATH_TO_PROXY_CERTIFICATE>` with the path to the proxy public certificate and `<PATH_TO_PROXY_CERTIFICATE_KEY>` with the path to the private key.
```

        proxy_ssl on;
        proxy_pass dbm-metrics-intake.{{< region-param key="dd_site" >}}:443;
    }
    server {
        listen 3840 ssl; #listen for database monitoring samples
        proxy_ssl_verify on;
        proxy_ssl on;
        proxy_pass dbquery-intake.{{< region-param key="dd_site" >}}:443;
    }
    server {
        listen 3841 ssl; #listen for network devices metadata
        proxy_ssl_verify on;
        proxy_ssl on;
        proxy_pass ndm-intake.{{< region-param key="dd_site" >}}:443;
    } 
    server {
        listen 3842 ssl; #listen for network devices traps
        proxy_ssl_verify on;
        proxy_ssl on;
        proxy_pass snmp-traps-intake.{{< region-param key="dd_site" >}}:443;
    }
    server {
        listen 3843 ssl; #listen for instrumentations telemetry data
        proxy_ssl_verify on;
        proxy_ssl on;
        proxy_pass instrumentation-telemetry-intake.{{< region-param key="dd_site" >}}:443;
    }
    server {
        listen 3844 ssl; #listen for appsec events (deprecated)
        proxy_ssl_verify on;
        proxy_ssl on;
        proxy_pass appsecevts-intake.{{< region-param key="dd_site" >}}:443;
    }
}
```
{{% /tab %}}
{{< /tabs >}}

> **Note**: Download the certificate with one of the following commands:

Reviewer: ```suggestion
**Note**: Install the common root CA certificates with one of the following commands:
```

Reviewer 2: Move this section above the config

```shell
sudo apt-get install ca-certificates # (Debian, Ubuntu)
yum install ca-certificates # (CentOS, Red Hat)
```

The file might be located at `/etc/ssl/certs/ca-certificates.crt` for Debian, Ubuntu or `/etc/ssl/certs/ca-bundle.crt` for CentOS, Red Hat.

**Note**: You can remove `proxy_ssl_verify on` if you are unable to get the certificates on the proxy host, but be aware that NGINX will not be able to verify Datadog's intake certificate in that case.

#### Datadog Agent configuration

Edit each Agent configuration file to point to Nginx by setting its `dd_url` to the address of Nginx, for example: `nginx.example.com`.
This `dd_url` setting can be found in the `datadog.yaml` file.

`dd_url: "<SCHEME>://nginx.example.com:3834"`

> Replace `<SCHEME>` by `https` if you previously chose the NGINX `HTTPS` configuration, or by `http` otherwise.

Reviewer 2: ```suggestion
Replace `<SCHEME>` with `https` if you previously chose the HAProxy HTTPS configuration, or with `http` if you did not choose HTTPS.
```

        logs_dd_url: nginx.example.com:3839
        # Comment the line below to use encryption between the Agent and NGINX
        logs_no_ssl: true
    activity:
        logs_dd_url: nginx.example.com:3839
        # Comment the line below to use encryption between the Agent and NGINX
        logs_no_ssl: true
    samples:
        logs_dd_url: nginx.example.com:3840
        # Comment the line below to use encryption between the Agent and NGINX
        logs_no_ssl: true

network_devices:
    metadata:
        logs_dd_url: nginx.example.com:3841
        # Comment the line below to use encryption between the Agent and NGINX
        logs_no_ssl: true
    snmp_traps:
        forwarder:
            logs_dd_url: nginx.example.com:3842
            # Comment the line below to use encryption between the Agent and NGINX
            logs_no_ssl: true

appsec_config (deprecated):
    appsec_dd_url: haproxy.example.com:3844

```

When using encryption between the Agent and NGINX, if the Agent does not have access to the proxy certificate, is unable to validate it, or the validation is not needed, you can edit the `datadog.yaml` Agent configuration file and set `skip_ssl_validation` to `true`.

> With this option set to `true`, the Agent will skip the certificate validation step so it will not verify the identity of the proxy but the communication will still be encrypted with SSL/TLS.

Reviewer 2: ```suggestion
With this option set to `true`, the Agent skips the certificate validation step and does not verify the identity of the proxy, but the communication is still encrypted with SSL/TLS.
```

```yaml
skip_ssl_validation: true
```

> When sending logs over TCP, see <a href="/agent/logs/proxy">TCP Proxy for Logs</a>.

Reviewer 2: ```suggestion
When sending logs over TCP, see [TCP Proxy for Logs][5].
```

{{< partial name="whats-next/whats-next.html" >}}

[1]: http://haproxy.1wt.eu
[2]: http://www.haproxy.org/#perf
[3]: https://www.haproxy.com/blog/haproxy-ssl-termination/

> [4]: https://www.nginx.com

Reviewer 2: ```suggestion
[4]: https://www.nginx.com
[5]: /agent/logs/proxy
```
